Check if html-build is up to date before building

foolip opened the pull request:
Fixes https://github.com/whatwg/html-build/issues/53

domenic commented on build.sh:
```diff
@@ -57,6 +57,24 @@ do
   esac
 done
 
+if [ "$DO_UPDATE" == true ] && [ "$QUIET" == false ]; then
```
I don't think we want to skip the whole check on $QUIET; we just want to skip the "Checking if html-build is up to date..." line

foolip replied:
You're right, at the very least `exit 1` should run even if quite.

domenic replied:
In general I am pretty sure quiet is meant to mean "no output in the success case", but in the error case we still want nice error messages.

domenic commented on build.sh:
```diff
@@ -57,6 +57,24 @@ do
+if [ "$DO_UPDATE" == true ] && [ "$QUIET" == false ]; then
+  echo "Checking if html-build is up to date..."
+  ORIGIN_URL=$(git remote get-url origin)
+  git fetch -q $ORIGIN_URL master
```
I think -q should only be if $QUIET is set. It seems important to know what is happening to my local git checkout.

foolip replied:
Because this is fetching $ORIGIN_UR` and not origin, it doesn't update any refs other than FETCH_HEAD. It will print something like this:
```
From github.com:whatwg/html-build
 * branch            master     -> FETCH_HEAD
```

I think this is useless to show by default, but we should omit the -q if VERBOSE I think. (Will update when we agree on behavior.)

If you're concerned about changing FETCH_HEAD, we could use a hidden ref like bisect and rerere does, perhaps refs/html-build/origin/master?

domenic replied:
I'm not sure I know enough to be concerned about changing FETCH_HEAD :). What does that do?

If this is indeed meant to not mess with origin, then it seems OK to be hiding it behind verbose and silencing it by default. It doesn't output any more useful information, like number of commits in the update, or a progress bar?

foolip replied:
FETCH_HEAD is simply the place where the ref ends up if you specify both a remote/url and a branch, but no target branch as in `git fetch origin master:newbranch`. I'm not worried about stomping on it, if people have done a fetch like this and then not immediately used FETCH_HEAD, then they probably won't mind us overwriting it.

Even passing -v to git fetch doesn't show any interesting information I'm afraid, it's still like the above pasted.

domenic replied:
Even if there are new commits? Kind of a shame you don't get to see how behind you are.

But yeah, OK, sounds good to me.

foolip replied:
We could print the new commits with `git log HEAD..FETCH_HEAD` and I tried that, but I'm not sure what a contributor is supposed to do with that information.

domenic replied:
Eh, I guess it's fine to leave it for them to discover on their own when they do the pull or whatever. I was thinking it might be nice to explain why we're erroring out---what critical functionality have you missed out on. But no big deal.

foolip replied:
What I could easily do is print how many commits behind you are. A shortlog would also be easy, of course. I'll push two fixups, one with each improvement, and you can decide if you like either or both of them.

foolip commented:
About $QUIET, how much do we want that to print in case of failure? It's documented as "Don't emit any messages except errors/warnings" but actually prints nothing at all in case wattsi fails. That was deliberate in f7bbfdbe6801c1cc501a24f33d1826956bbd54fa but perhaps not a good idea. I'd like to fix that and be similarly verbose with this failure.

domenic commented:
I think the full error message is fine. The use of -q is for the Unixey "Rule of Silence" in the success case. It would indeed be great to fix that change as well.

foolip commented:
Fixing that in https://github.com/whatwg/html-build/pull/73

foolip commented:
OK, pushed a fixup that I think addresses all comments. It's on a branch deliberately not rebase on master so that you can test it locally before merging.

foolip commented:
OK, now you get to pick. I kind of like commit count + shortlog, in the end.

domenic commented:
This looks lovely. I do like the commit count + shortlog. We have the best build scripts ever! Despite them being crazy cobbled-together bash scripts ^_^

foolip commented:
Should I merge, or do you want to do some more testing?

domenic commented:
Go for it!
